### ISSUE TITLE
en_us.POSIX locale environment for .NET Core on Linux

### ISSUE THREAD

**Issue by alexkeh** [dotnet/corefx#28611]
Does .NET Core support Linux in en_us.POSIX locale environment?

Based on [this commit](https://github.com/dotnet/corefx/commit/513efd86d5fb290098d6823b7b847cf50dd01951), which adds a regression test for a bug fix and it invokes CultureInfo("en-US-POSIX"). However, my team has run into a fundamental problem running .NET Core tests in Linux with en_us.POSIX locale. We have no issue with the en_us.UTF8 locale. For example, string.Compare() does not work properly in case-sensitive comparison.

- string.Compare("C03DATE", "C03DATE", true) return zero, which is expected
- string.Compare("c03DATE", "C03DATE", true) return non-zero, which is not expected because it is case-insensitive comparison. Third parameter is true.  In other locales, this method returns zero. 

Thanks to @divega for providing some initial help on this question.

**Comment by tarekgh** [comment]
@mwoo-o The issue here is if .Net Core supporting POSIX. the issue here is what is the POSIX string comparison behavior. if you look at the link I sent before https://marc.info/?l=icu&m=101779972120281&w=2 you'll see 

```
The purpose of the en_US_POSIX locale in general is to make ICU more locale insensitive in certain circumstances.
```

This makes the POSIX locale has special behavior for string comparison. in another word, en_US_POSIX behavior is really different than en_US behavior. the string comparison behavior is defined by LCDR (Unicode.org) here https://unicode.org/cldr/trac/browser/trunk/common/collation/en_US_POSIX.xml and this behavior is carried by ICU library which the framework depends on for performing the string comparison operations. so, the framework is not really defining the behavior here. 

You still can control the behavior here by setting the current culture to whatever locale you want or even Invariant if you like.

**Comment by mwoo-o** [comment]
We can't control the locale that are used by the customers.  So that will mean that if we want to
support POSIX, then we will have to make some code differentiation for POSIX  locale in our driver. 
Is this correct understanding?

Questions:
Does Windows support POSIX locale?
Other than string functions in POSIX, is there something that we need to aware for POSIX behavior?
Any link information for POSIX behaviors will be appreciated.

**Comment by tarekgh** [comment]
> Does Windows support POSIX locale?

No 

> Other than string functions in POSIX, is there something that we need to aware for POSIX behavior?
> Any link information for POSIX behaviors will be appreciated.
.Net Core is just using ICU library. you may consult the ICU library documentation here http://userguide.icu-project.org/locale

> We can't control the locale that are used by the customers. 

You can still set the current culture in your code. using CultureInfo.CurrentCulture

> So that will mean that if we want to support POSIX, then we will have to make some code differentiation for POSIX locale in our driver. Is this correct understanding?

We discourage users to set their locale to POSIX because of the issues we talked about here. If the user decided to use POSIX locale, then that is their choice and they will get the behavior for that locale which may not be desired but that is what the user chose.

**Comment by tarekgh** [comment]
@mwoo-o 

The .Net framework chose to use the current culture for the default comparisons. You have the choice to perform the comparisons differently as you desired. Just pass the parameter StringComparison.OrdinalIgnoreCase or StringComparison.InvariantCultureIgnoreCase to the comparison APIs and you'll get what you want:

https://docs.microsoft.com/en-us/dotnet/api/system.string.equals?view=netframework-4.7.1#System_String_Equals_System_String_System_StringComparison_
https://docs.microsoft.com/en-us/dotnet/api/system.string.equals?view=netframework-4.7.1#System_String_Equals_System_String_System_String_System_StringComparison_
https://docs.microsoft.com/en-us/dotnet/api/system.string.compare?view=netframework-4.7.1#System_String_Compare_System_String_System_String_System_StringComparison_

why you can't do that?

**Comment by mwoo-o** [comment]
@tarekgh 
Are you suggesting to use
string.Compare(string1, string2, StringComparison. InvariantCultureIgnoreCase)?

In our code, we have a lot of 
 string.Compare(string1, string2, true)
If we change it to use StringComparison. InvariantCultureIgnoreCase, do you see any impact on other
locales that will cause different behaviors?

**Comment by tarekgh** [comment]
@mwoo-o 

You need to look at the places you are doing string comparison in your code and figure out if the comparison you are doing is one of the following cases:
- Does the comparison need to be performed linguistically according to the current culture (whatever the user choose the current culture)? if so, then you don't need to pass the StringComparison at all and we'll use the current culture to perform the comparison. example of that, you want to comparison handle the Turkish I comparison correctly when the user chooses to set the current culture as Turkish locale. This type of comparisons usually needed for the strings used by the UI.
- Does the comparison need to be performed ordinal and not Linguistically, at that time just pass the StringComparison.OrdinalIgnoreCase. this comparison will do the comparison as comparing every character in the first string to another character in the second string (and handling the casing according to Unicode spec). example of that, if parsing XML element and you know the element names are usually in English. Or comparing file names and paths.
- Does the comparison need to be performed linguistically but independently from the current culture, then pass StringComparison.InvariantCultureIgnoreCase. Note that, if the user sets the current locale to Turkish and comparing strings containing Turkish I then the comparison may not perform as the user expect because you didn't choose to use the current culture. using Invariant comparison is useful when want to have consistent behavior regardless of the linguistic correctness according to current culture.
- Does the comparison need to behave according to a specific language, if so then you pass the culture has this language to the string comparison method? for example, if you want all comparisons performed according to the English language, then you pass en-US culture to the Compare method. 

You may look at the code sample https://docs.microsoft.com/en-us/dotnet/api/system.stringcomparison?view=netframework-4.7.1#examples to get some idea about what you expect from each comparison type.

To summarize, you need to know what behavior you want the strings compared with and choose the right option. if you have some example from your code, I can help you telling which option might be good in that case.

**Comment by mwoo-o** [comment]
@tarekgh 
Thank you very much for the information. We will need to go through 100+ string.Compare() calls. Will let you know if we have questions.

**Comment by tarekgh** [comment]
https://github.com/dotnet/docs/issues/8179
